src/streamed/state.ts

```diff
@@ -165,135 +167,258 @@ export class AuctionState {
+    // TODO(nlordell): We need to pre-fetch the batch ID based on the block
+    // number for this event to be able to accurately set this value.
+    this.order(owner, parseInt(id)).validUntil = null
+  }
 
+  /**
+   * Applies an order deletion event to the auction state.
+   */
+  private applyOrderDeletion({ owner, id }: Event<BatchExchange, "OrderDeletion">) {
+    const order = this.order(owner, parseInt(id))
+    order.buyToken = 0
+    order.sellToken = 0
+    order.validFrom = 0
+    order.validUntil = 0
+    order.priceNumerator = BigInt(0)
+    order.priceDenominator = BigInt(0)
+    order.remainingAmount = BigInt(0)
+  }
+
+  /**
+   * Applies an order placement event to the auction state.
+   *
+   * @throws
+   * In strict mode, throws if the order ID from the event does not match the
+   * expected order ID based on the number of previously placed orders.
+   */
+  private applyOrderPlacement(order: Event<BatchExchange, "OrderPlacement">) {
+    const userOrders = this.account(order.owner).orders
+    const orderId = userOrders.length
     if (this.options.strict) {
-      assertAccountsAreValid(this.lastBlock, this.accounts.entries())
+      assert(
+        orderId == parseInt(order.index),
+        `user ${order.owner} order ${order.index} added as the ${orderId}th order`,
+      )
     }
+
+    userOrders.push({
+      buyToken: parseInt(order.buyToken),
+      sellToken: parseInt(order.sellToken),
+      validFrom: parseInt(order.validFrom),
+      validUntil: parseInt(order.validUntil),
+      priceNumerator: BigInt(order.priceNumerator),
+      priceDenominator: BigInt(order.priceDenominator),
+      remainingAmount: BigInt(order.priceDenominator),
+    })
   }
 
   /**
-   * Adds a token to the account state.
+   * Applies an emulated solution reversion event.
+   */
+  private applySolutionReversion(
+    { submitter, burntFees }: Event<BatchExchange, "SolutionSubmission">,
+  ) {
+    this.updateBalance(submitter, 0, amount => amount - BigInt(burntFees))
+  }
+
+  /**
+   * Applies a solution submission event to the auction state.
    *
-   * @throws If a token with the same ID has already been added.
+   * @throws
+   * In strict mode, throws if the account balances are left in an invalid state
+   * while applying a solution. This check has to be done after applying all
+   * trades, as a user balance can temporarily go below 0 when a solution is
+   * being applied.
+   */
+  private applySolutionSubmission(
+    { submitter, burntFees }: Event<BatchExchange, "SolutionSubmission">,
+  ) {
+    this.updateBalance(submitter, 0, amount => amount + BigInt(burntFees))
+    if (this.options.strict) {
+      for (const [user, { balances }] of this.accounts.entries()) {
+        for (const [token, balance] of balances.entries()) {
+          assert(balance >= BigInt(0), `user ${user} token ${token} balance is negative`)
+        }
+      }
+    }
```

Comment:
https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L209-L212

and

https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L478-L485